hpx/hpx_distributed.cc

```diff
@@ -0,0 +1,369 @@
+
+      for (int r = 0; r < n_ranks; ++r) {
+        long r_first_point = r * graph.max_width / n_ranks;
+        long r_last_point = (r + 1) * graph.max_width / n_ranks - 1;
+        for (long p = r_first_point; p <= r_last_point; ++p) {
+          rank_by_point[p] = r;
+          tag_bits_by_point[p] = p - r_first_point;
+          assert((tag_bits_by_point[p] & ~0x7F) == 0);
+        }
+      }
+
+      long max_deps = 0;
+      for (long dset = 0; dset < graph.max_dependence_sets(); ++dset) {
+        for (long point = first_point; point <= last_point; ++point) {
+          long deps = 0;
+          for (auto interval : graph.dependencies(dset, point)) {
+            deps += interval.second - interval.first + 1;
+          }
+          max_deps = std::max(max_deps, deps);
+        }
+      }
+
+      // Create input and output buffers.
+      std::vector<std::vector<std::vector<char> > > inputs(n_points);
+      std::vector<std::vector<const char *> > input_ptr(n_points);
+      std::vector<std::vector<size_t> > input_bytes(n_points);
+      std::vector<long> n_inputs(n_points);
+      std::vector<std::vector<char> > outputs(n_points);
+      std::vector<std::vector<char> > outputs_new(n_points);
+
+      for (long point = first_point; point <= last_point; ++point) {
+        long point_index = point - first_point;
+
+        auto &point_inputs = inputs[point_index];
+        auto &point_input_ptr = input_ptr[point_index];
+        auto &point_input_bytes = input_bytes[point_index];
+
+        point_inputs.resize(max_deps);
+        point_input_ptr.resize(max_deps);
+        point_input_bytes.resize(max_deps);
+
+        for (long dep = 0; dep < max_deps; ++dep) {
+          point_inputs[dep].resize(graph.output_bytes_per_task);
+          point_input_ptr[dep] = point_inputs[dep].data();
+          point_input_bytes[dep] = point_inputs[dep].size();
+        }
+
+        auto &point_outputs = outputs[point_index];
+        point_outputs.resize(graph.output_bytes_per_task);
+
+        auto &point_outputs_new = outputs_new[point_index];
+        point_outputs_new.resize(graph.output_bytes_per_task);
+      }
+
+      // Cache dependencies.
+      std::vector<std::vector<std::vector<std::pair<long, long> > > >
+          dependencies(graph.max_dependence_sets());
+      std::vector<std::vector<std::vector<std::pair<long, long> > > >
+          reverse_dependencies(graph.max_dependence_sets());
+      for (long dset = 0; dset < graph.max_dependence_sets(); ++dset) {
+        dependencies[dset].resize(n_points);
+        reverse_dependencies[dset].resize(n_points);
+
+        for (long point = first_point; point <= last_point; ++point) {
+          long point_index = point - first_point;
+
+          dependencies[dset][point_index] = graph.dependencies(dset, point);
+          reverse_dependencies[dset][point_index] =
+              graph.reverse_dependencies(dset, point);
+        }
+      }
+      
+      // optimization for certain types
```

Comment:
Can you expand this comment a bit more to say what the optimization is and what the requirements are for it to be applicable?

E.g., if it's something like that the offset is always 0 and width is always graph.max_width, then that's something you could check directly instead of hard-coding the set of dependence patterns that satify that (which may hypothetically be expanded in the future).

(And also, what happens in the path where this optimization is *not* taken? Looks like it just uses a regular for loop, but what does that mean? Less opportunity for overlap? And if so, how?)

Reply:
It is related to the dependence types. The optimization works for a type if only the boundary cores of a node need to communicate with cores from another nodes, and the inner cores only use on-node sharing; or a type does not involve communication.

If there are new dependence patterns added in the future, I could certainly work on it.

The optimization is using one hpx for loop to do the communication and execute kernel. The cores that does not have communication can proceed without block. The non-optimization is using two hpx for loop, one is to do communication and wait for all communication is done, then a second for loop to execute kernel.